seaborn/_core/data.py

```diff
@@ -258,3 +265,26 @@ def _assign_variables(
         frame = pd.DataFrame(plot_data)
 
         return frame, names, ids
+
+
+def convert_dataframe_to_pandas(data: object) -> pd.DataFrame:
+    """Use the DataFrame exchange protocol, or fail gracefully."""
+    if isinstance(data, pd.DataFrame):
+        return data
+    if not hasattr(pd.api, "interchange"):
+        msg = (
+            "Support for non-pandas DataFrame objects requires a version of pandas"
+            "that implements the DataFrame interchange protocol. Please upgrade "
+            "your pandas version or coerce your data to pandas before passing "
+            "it to seaborn."
+        )
+        raise TypeError(msg)
+
+    try:
+        return pd.api.interchange.from_dataframe(data)
```

Comment:
just sharing this plotly PR for reference: https://github.com/plotly/plotly.py/pull/4286/files

if it's possible to tell in advance which columns need converting to pandas, then it may be more efficient only select them before hand using `__dataframe__().select_columns_by_name`

Reply:
In general that is somewhere between difficult and hard. For instance, as currently written, when you do `Plot(data).add(so.Dots(), x="x", y="y')`, the data interchange will happen when the `Plot()` constructor executes but the columns that are needed from the constructor aren't known until `Plot.add()` executes.

I've added a comment to the code noting that this is a known inefficiency but I don't think there's much that can be done about it save a large rewrite of how data ingest works so I am going to say we need to live with it for now.

Reply:
sounds good, thanks for explaining!